File: Graph/ts/bfs.ts
import { Graph } from "./types";

export function breadthFirstTraversal(graph: Graph, source: string): string[] {
  if (!source) return [];

  const order: string[] = [];
  const queue: string[] = [source];
  const visited = new Set<string>([source]);

  while (queue.length > 0) {
    const current = queue.shift()!;
    order.push(current);

    for (const neighbor of graph[current] ?? []) {
      if (visited.has(neighbor)) continue;
      visited.add(neighbor);
      queue.push(neighbor);
    }
  }

  return order;
}

File: Graph/go/bfs.go
package graph

// BreadthFirstTraversal returns the order of nodes visited during a BFS.
func BreadthFirstTraversal(g Graph, source string) []string {
	if source == "" {
		return []string{}
	}

	order := []string{}
	queue := []string{source}
	visited := map[string]struct{}{source: {}}

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		order = append(order, node)

		for _, neighbor := range g[node] {
			if _, seen := visited[neighbor]; seen {
				continue
			}
			visited[neighbor] = struct{}{}
			queue = append(queue, neighbor)
		}
	}

	return order
}
